### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] False positive in 8.59.0 with mapped conditional class method

### ISSUE THREAD

**Issue by joshkel** [typescript-eslint/typescript-eslint#12275]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgKIDcLgM7IN4BQyycAFAB4Bcy2YUoA5gJTXoD2wAJgNxHIBGFaiACuAW37QWydl14BfAgTABPAA4oAglAbYAPAGlkEcpBCdcAawgq2MNJhwA%2BZAF58fANoAFZKGTWtvYYWGDYALrUITg%2B4camWBbIpAB0aXDUoDDQyJpMbi6ynMgA-LnIwhCYUAqeBuG8BJwQCAA2cFAobXDYuADCHsQQYsBghvFmSYF2DqHYTqRVodQGADTIaSkdutTauoZO0kUKSs3dncgIbCC0l9R9vGftF1c3YAE21ADkcF-IAD7IL78L6PFrPLrXW7bai0eggBgA5CiCTQRoAenRaAAygAZUDvbBwFS4HrYaBgYDXPy4EQgEAtCC9Doqdb8ETvTpiNjoRh%2Bd4IOAicm4AAq2IATABmAAsAFYCAgUsNRqRAuttiRSSAVExeEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
interface Events {
  a(x: string): void;
  b(x: number): void;
}

type Args<K extends keyof Events> = {
  [P in keyof Events]: Events[P] extends (...a: infer A) => void ? A : never;
}[K];

declare class C {
  emit<K extends keyof Events>(event: K, ...args: Args<K>): void;
}

declare const c: C;
declare const key: 'a' | 'b';
declare const arg: string | number;

// ESLint says assertion is unnecessary, but removing it causes TS2345
c.emit(key, arg as any);
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No ESLint warning, because a cast is necessary to make TypeScript happy.

### Actual Result

"This assertion is unnecessary since the receiver accepts the original type of the expression."

**Comment by bradzacher** [comment]
This is the exact same case as https://github.com/typescript-eslint/typescript-eslint/issues/12277 

The type of the function arguments is a discriminated union just like the object case from #12277.
